mpi: autogen MPI functions - pt2pt

hzhou opened the pull request:
## Pull Request Description

Auto generate MPI functions based on API scraped from mpi-standard. 

Using script to generate top-level code avoids manually maintain bolierplate code and provides consistency such as automatic api documentation and parameter validations.

The base file, `maint/mpi_standard_api.txt`, is the vanilla base from official mpi-standard repository, it does not require any manual maintenance.

Additional customization file, e.g. `src/mpi/pt2pt/mpi_pt2pt_api.txt`, allows for `MPICH` specific parts, such as body of implementations, chimp-specific man-page notes. Ideally, we will keep these files concise.

## What is being generated
### 1. Profiling block
```
/* -- Begin Profiling Symbol Block for routine MPI_Send */
#if defined(HAVE_PRAGMA_WEAK)
#pragma weak MPI_Send = PMPI_Send
#elif defined(HAVE_PRAGMA_HP_SEC_DEF)
#pragma _HP_SECONDARY_DEF PMPI_Send  MPI_Send
#elif defined(HAVE_PRAGMA_CRI_DUP)
#pragma _CRI duplicate MPI_Send as PMPI_Send
#elif defined(HAVE_WEAK_ATTRIBUTE)
int MPI_Send(const void *buf, int count, MPI_Datatype datatype, int dest, int tag, MPI_Comm comm) __attribute__ ((weak, alias("PMPI_Send")));
#endif
/* -- End Profiling Symbol Block */

/* Define MPICH_MPI_FROM_PMPI if weak symbols are not supported to build
   the MPI routines */
#ifndef MPICH_MPI_FROM_PMPI
#undef MPI_Send
#define MPI_Send PMPI_Send

#endif

```
One hundred percent boilerplate.
### 2. Man page
```
/*@
   MPI_Send - Performs a blocking send

Input Parameters:
+ buf - initial address of send buffer (choice)
. count - number of elements in send buffer (non-negative integer)
. datatype - datatype of each send buffer element (handle)
. dest - rank of destination (non-negative integer)
. tag - message tag (integer)
- comm - communicator (handle)

Notes:
This routine may block until the message is received by the destination
process.
.N ThreadSafe

.N Fortran

.N Errors
.N MPI_SUCCESS
.N MPI_ERR_BUFFER
.N MPI_ERR_COMM
.N MPI_ERR_COUNT
.N MPI_ERR_RANK
.N MPI_ERR_TAG
.N MPI_ERR_TYPE

.seealso: MPI_Isend, MPI_Bsend
@*/

```
Parameters are exactly consistent to mpi-standard.

Notes are separately maintained in, e.g. `src/mpi/pt2pt/mpi_pt2pt_api.txt`, among other things.

Now the error code documentation is synchronized with the validations below.

### 3. Parameter validations
```
#ifdef HAVE_ERROR_CHECKING
    {
        MPID_BEGIN_ERROR_CHECKS;
        {
            MPIR_ERRTEST_COMM(comm, mpi_errno);
        }
        MPID_END_ERROR_CHECKS;
    }
#endif /* HAVE_ERROR_CHECKING */
    MPIR_Comm_get_ptr(comm, comm_ptr);

#ifdef HAVE_ERROR_CHECKING
    {
        MPID_BEGIN_ERROR_CHECKS;
        {
            MPIR_ERRTEST_USERBUFFER(buf, count, datatype, mpi_errno);
            MPIR_ERRTEST_COUNT(count, mpi_errno);
            MPIR_ERRTEST_DATATYPE(datatype, "datatype", mpi_errno);
            if (!HANDLE_IS_BUILTIN(datatype)) {
                MPIR_Datatype *datatype_ptr = NULL;
                MPIR_Datatype_get_ptr(datatype, datatype_ptr);
                MPIR_Datatype_valid_ptr(datatype_ptr, mpi_errno);
                if (mpi_errno)
                    goto fn_fail;
                MPIR_Datatype_committed_ptr(datatype_ptr, mpi_errno);
                if (mpi_errno)
                    goto fn_fail;
            }
            MPIR_ERRTEST_SEND_RANK(comm_ptr, dest, mpi_errno);
            MPIR_ERRTEST_SEND_TAG(tag, mpi_errno);
            MPIR_Comm_valid_ptr(comm_ptr, mpi_errno, FALSE);
            if (mpi_errno)
                goto fn_fail;
        }
        MPID_END_ERROR_CHECKS;
    }
#endif /* HAVE_ERROR_CHECKING */

```
This is typically more consistent and complete than current code base.
 
### 4. Early returns
```
    /* Return immediately for dummy process */
    if (unlikely(dest == MPI_PROC_NULL)) {
        goto fn_exit;
    }
```
Maybe this should manually maintained with the "body of routines" -- together in e.g. `src/mpi/pt2pt/mpi_pt2pt_api.txt` (or a different file if we want it to be separate from documentations)

### 5. Error return
```
  fn_exit:
    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPI_SEND);
    MPID_THREAD_CS_EXIT(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
    return mpi_errno;
  fn_fail:
    /* --BEGIN ERROR HANDLINE-- */
#ifdef HAVE_ERROR_CHECKING
    {
        mpi_errno =
            MPIR_Err_create_code(mpi_errno, MPIR_ERR_RECOVERABLE, __func__, __LINE__, MPI_ERR_OTHER,
                                 "**mpi_send", "**mpi_send %p %d %D %i %t %C", buf, count, datatype,
                                 dest, tag, comm);
    }
#endif
    mpi_errno = MPIR_Err_return_comm(comm_ptr, __func__, mpi_errno);
    /* --END ERROR HANDLING-- */
    goto fn_exit;
```
The special argument format is now more consistent than current code base.

## What is not being generated
Mostly, "body of routines" and special documentation notes, and some special directives if any, maintained in, e.g. `src/mpi/pt2pt/mpi_pt2pt_api.txt`.
```
MPI_Send:
    .file: pt2pt/send.c
    .desc: Performs a blocking send
    .seealso: MPI_Isend, MPI_Bsend
/*
    Notes:
    This routine may block until the message is received by the destination
    process.
*/
{
    MPIR_Request *request_ptr = NULL;

    mpi_errno = MPID_Send(buf, count, datatype, dest, tag, comm_ptr,
                          MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
    if (mpi_errno != MPI_SUCCESS)
        goto fn_fail;

    if (request_ptr == NULL) {
        goto fn_exit;
    }

    mpi_errno = MPID_Wait(request_ptr, MPI_STATUS_IGNORE);
    MPIR_ERR_CHECK(mpi_errno);

    mpi_errno = request_ptr->status.MPI_ERROR;
    MPIR_Request_free(request_ptr);

    if (mpi_errno != MPI_SUCCESS)
        goto fn_fail;
}
```
There are still quite some repetitive parts in them, so we could simplify it by notes like --
```
    .body_of_routine: MPID_Send, MPID_Wait /* or MPIR */
```
The parameters and error check etc. can be easily generated.

However, I am not sure whether that is worthwhile yet, so I am leaving that to-do separately.

## Expected Impact
In this iteration, we try to minimize differences between the generated results from old (now removed) source code. There should be no difference for the body of routines. Some differences are produced in the documentations, parameter validations, and short comments and some coding styles. These are mostly due to inconsistency in the old code. Nevertheless, we should manually examine the differences to ensure the generated results are good.

This work should make future work such as the "embiggenment" and adding new API much easier.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
[pt2pt-diff.txt](https://github.com/pmodels/mpich/files/5411228/pt2pt-diff.txt)

hzhou commented:
[pt2pt-diff.txt](https://github.com/pmodels/mpich/files/5411271/pt2pt-diff.txt)

wesbland commented:
We're writing new scripts in Perl? I'm really not excited to have to re-remember Perl to migrate all of my #4715 work to this. :(

hzhou commented:
> We're writing new scripts in Perl? I'm really not excited to have to re-remember Perl to migrate all of my #4715 work to this. :(

I can easily make them in Python. Is that preferred?

wesbland commented:
> I can easily make them in Python. Is that preferred?

I thought we'd agreed that there'd be no new Perl, but I can't remember for sure now. It's certainly my preference.

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/warnings/auto

hzhou commented:
test:mpich/warnings/auto

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
In stead of manually rewrite the script from Perl to Python, I invested time to write a `perl_to_python` script in stead. The results seem working. I can prototype in Perl and publish in Python.

hzhou commented:
test:mpich/ch4/ofi

wesbland commented:
> In stead of manually rewrite the script from Perl to Python, I invested time to write a `perl_to_python` script in stead. The results seem working. I can prototype in Perl and publish in Python.

Technically that's true. In practice, that script is far from nice. 😄 

I'm fine with continuing in Perl until it's time for the final version and then moving everything over if that's how you want to do it.

hzhou commented:
>  In practice, that script is far from nice. smile

So there is space to improve :) Pointers?

wesbland commented:
I just mean that the Pythonization of the Perl script still looks pretty machine generated. There are a number of obvious machine-generated variable names and concatenations that are common for this sort of thing.

hzhou commented:
test:mpich/authorship

hzhou commented:
test:mpich/custom

hzhou commented:
test:mpich/custom

hzhou commented:
test:mpich/custom

hzhou commented:
test:mpich/ch3/most

hzhou commented:
test:mpich/ch3/sock
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/compilers

hzhou commented:
test:mpich/ch3/compilers

hzhou commented:
test:mpich/ch3/sock

hzhou commented:
test:mpich/ch3/sock

hzhou commented:
test:mpich/ch3/tcp

hzhou commented:
test:mpich/ch3/sock

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
@wesbland I have refactored the scripts into python packages and modules so that the code can be easily shared across different scripts. I have transitioned into directly updating python scripts at this point. Would you like to take another review?

@raffenet This PR is ready for your review.

hzhou commented:
test:mpich/ch3/sock

raffenet commented on autogen.sh:
```diff
@@ -601,6 +601,27 @@ else
+## Check for Python 3
+########################################################################
+
+echo_n "Checking for Python 3... "
```
For the curious: what Python 3 features are we relying on?

hzhou replied:
I haven't really checked. But since Python2 is deprecated there is no reason to write new scripts for Python2. By requiring Python 3, we are more future proof and have less burden of support.

Good catch for the doubled output, I'll fix it.

raffenet replied:
> I haven't really checked. But since Python2 is deprecated there is no reason to write new scripts for Python2. By requiring Python 3, we are more future proof and have less burden of support.

Yes, but Python 2 vs. 3 on today's systems is not as straight-forward as one would like. My preference would be to make the script compatible with either. But if that is too much burden, requiring 3 is better. That would be consistent with what Yaksa already does anyway.

hzhou replied:
> My preference would be to make the script compatible with either. But if that is too much burden,...

It is not too much burden now, but it is a burden that, once in, we are going to carry for a long time; not only for this script, but for every new scripts in order to keep the policy consistency. This is autogen, so I think it is less of an issue. But we should discuss on the general policy forward. I can easily make the revision to reflect whatever we decide then.

raffenet commented on src/mpi/errhan/errnames.txt:
```diff
@@ -1409,9 +1409,9 @@ is too big (> MPIU_SHMW_GHND_SZ)
 ## MPI-3 Mprobe routines
 **mpi_improbe: MPI_Improbe failed
-**mpi_improbe %d %d %C %p %p %p: MPI_Improbe(source=%d, tag=%d, comm=%C, flag=%p, message=%p, status=%p)
+**mpi_improbe %i %t %C %p %p %p: MPI_Improbe(source=%i, tag=%t, comm=%C, flag=%p, message=%p, status=%p)
```
Commit subject says mrecv and imrecv, but changes are for mprobe and improbe.

hzhou replied:
I'll fix it.

raffenet reviewed (commented):
This looks pretty nice. I might squash the addition and refactoring of the generation script. Also note that line spacing in the generated bodies of routines is doubled.
```
    /* ... body of routine ... */
        MPIR_Request *request_ptr = NULL;


        mpi_errno = MPID_Send(buf, count, datatype, dest, tag, comm_ptr,

                              MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);

        if (mpi_errno != MPI_SUCCESS)

            goto fn_fail;


        if (request_ptr == NULL) {

            goto fn_exit;

        }


        mpi_errno = MPID_Wait(request_ptr, MPI_STATUS_IGNORE);

        MPIR_ERR_CHECK(mpi_errno);


        mpi_errno = request_ptr->status.MPI_ERROR;

        MPIR_Request_free(request_ptr);


        if (mpi_errno != MPI_SUCCESS)

            goto fn_fail;

    /* ... end of body of routine ... */
```

wesbland commented on src/binding/c/pt2pt_api.txt:
```diff
@@ -0,0 +1,557 @@
+    mpi_errno = MPIR_Bsend_isend(buf, count, datatype, dest, tag, comm_ptr, NULL);
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+}
```
I don't understand the benefit here. Isn't this code that should be autogenerated too?

My understanding was that we'd autogenerate all of the error checking and then call the `MPIR` functions for everything else.

hzhou replied:
You are questioning the reason to allow direct code in the configuration file? It is for flexibility, so we have a back-door to customize certain functions when it is too complex or awkward to implement in the script or support new configure options.

In the case of `MPI_Bsend`, it is calling `MPIR_Bsend_isend` in stead of `MPIR_Bsend_impl` and having to wrap it with another `MPIR_Bsend_impl` seems too awkward :)

In addtion, the MPIR-impl functions are not inlined so for especially the pt2pt path, having direct code calling `MPID-` code may be desirable.

hzhou replied:
> Are you using something like f-strings? I think we use that in the MPI Standard repo and that's pretty new (3.8 I think?). If this is only a requirement for the developers, that's probably not a big deal, but if it's something that gets run at configure time, that might be harder.

No, I am using mostly '%' string -- no advanced features. The idea to require Python 3 is so we don't have to deal with any `Python 2` issues and since Python 2 is deprecated, all new scripts should target Python 3 anyway.

The scripts are only used during `autogen`, not `configure`, so I guess it is OK to be more picky in autogen. Not this, but we do have scripts in configure command, so I hope requiring Python 3 is not deal-breaker even there. (Here is another thing that reminds how great Perl is :) ).

raffenet replied:
I agree with @wesbland that some of the code here is probably better suited to autogeneration. But note that all the user argument checking **is** autogenerated, which is a big win IMO.

I am not opposed to the concept of the config file. We have this type of capability in `buildiface` scripts, too. Perhaps more aggressive generation doesn't need to happen in this first pass.

hzhou commented on maint/mpi_standard_api.txt:
```diff
@@ -0,0 +1,3419 @@
+APIs
```
Thanks. I was planning to wait for the PR to be merged and do an update, wasn't realize there is an `mpi-4-rc` branch. Is that branch nearly the final version -- API wise?

hzhou commented:
@raffenet The PR is updated with:
* Updated `mpi_standard_api.txt` with the latest standard repository
* Merged the `gen_binding_c.py` commits
* Fixed the double line output -- need `l = l.rstrip()` as `l.rstrip` does nothing. A leftover from Perl background.
* Rearranged some functions and added some comments/markers for readability
* Fixed the commit message for mprobe error messages.

hzhou commented:
test:mpich/ch4/ofi

raffenet commented on autogen.sh:
```diff
@@ -601,6 +601,27 @@ else
+fi
+
+if test -z "$PYTHON" -a 3 = `python3 -c 'import sys; print(sys.version_info[0])'`; then
+    PYTHON=python3
```
OK, now that I look at this closer I see that we are able to support both `python` and `python3`, which was most of my concern. I don't think that mess will be sorted out for some time.

raffenet commented on autogen.sh:
```diff
@@ -775,6 +775,13 @@ echo_n "Building ROMIO glue code... "
+########################################################################
+
+echo_n "generating MPI C functions..."
+$PYTHON maint/gen_binding_c.py
```
I was curious what would happen if I ran autogen at this commit, and I got an error that the `src/binding/c` directory didn't exist. I think we should consider merging this commit with the Makefile addition, and the actual replacement of pt2pt function with the autogenerated ones in order to consistently be able to build within the PR.

hzhou replied:
Ok, will do.

hzhou replied:
> I agree with @wesbland that some of the code here is probably better suited to autogeneration. But note that all the user argument checking is autogenerated, which is a big win IMO.

> I am not opposed to the concept of the config file. We have this type of capability in buildiface scripts, too. Perhaps more aggressive generation doesn't need to happen in this first pass.

There will be a lot of renaming and consistency refactoring to generate all the body of routines. It is fairly straight forward though. The mechanism is already there. For example, see https://github.com/pmodels/mpich/pull/4846/commits/cd92536769245af73b89b1c62d2c9e62929ab237#r533692730. Yeah, since it is not critical, I would prefer do it separately.

raffenet reviewed (commented):
A few more minor comments. I have no issue with the meat of the proposal at this point.

hzhou commented on src/binding/c/pt2pt_api.txt:
```diff
@@ -0,0 +1,557 @@
+MPI_Sendrecv:
+    .file: pt2pt/sendrecv.c
+    .desc: Sends and receives a message
+
```
Example where the body of routines is generated.

hzhou commented:
@raffenet Merged the commits as you mentioned. Also added `src/binding/c/README`, to provide some documentation on the mechanism as well as to ensure `src/binding/c` exist so the script won't fail. (or avoid extra error checking)

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

raffenet reviewed (commented):
One more nitpick - body of routine is indented one more level than it needs to be:
```
    /* Return immediately for dummy process */
    if (unlikely(dest == MPI_PROC_NULL)) {
        goto fn_exit;
    }

    /* ... body of routine ... */
        MPIR_Request *request_ptr = NULL;

        mpi_errno = MPID_Send(buf, count, datatype, dest, tag, comm_ptr,
                              MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
        if (mpi_errno != MPI_SUCCESS)
            goto fn_fail;

        if (request_ptr == NULL) {
            goto fn_exit;
        }

        mpi_errno = MPID_Wait(request_ptr, MPI_STATUS_IGNORE);
        MPIR_ERR_CHECK(mpi_errno);

        mpi_errno = request_ptr->status.MPI_ERROR;
        MPIR_Request_free(request_ptr);

        if (mpi_errno != MPI_SUCCESS)
            goto fn_fail;
    /* ... end of body of routine ... */
```

hzhou commented:
> 
> 
> One more nitpick - body of routine is indented one more level than it needs to be:
> 
> ```
>     /* Return immediately for dummy process */
>     if (unlikely(dest == MPI_PROC_NULL)) {
>         goto fn_exit;
>     }
> 
>     /* ... body of routine ... */
>         MPIR_Request *request_ptr = NULL;
> 
>         mpi_errno = MPID_Send(buf, count, datatype, dest, tag, comm_ptr,
>                               MPIR_CONTEXT_INTRA_PT2PT, &request_ptr);
>         if (mpi_errno != MPI_SUCCESS)
>             goto fn_fail;
> 
>         if (request_ptr == NULL) {
>             goto fn_exit;
>         }
> 
>         mpi_errno = MPID_Wait(request_ptr, MPI_STATUS_IGNORE);
>         MPIR_ERR_CHECK(mpi_errno);
> 
>         mpi_errno = request_ptr->status.MPI_ERROR;
>         MPIR_Request_free(request_ptr);
> 
>         if (mpi_errno != MPI_SUCCESS)
>             goto fn_fail;
>     /* ... end of body of routine ... */
> ```

Thanks. Now fixed.

hzhou commented:
test:mpich/ch4/ofi

raffenet reviewed (approved):
LGTM (pending tests).
